helpers/utils.py

```diff
@@ -0,0 +1,291 @@
+    :return clean:
+    """
+    flat_json_data = []
+
+    for data_item in data:
+        # use pandas 'nested_to_record' method to flatten json
+        # separate levels with '_'
+        flat_data_item = nested_to_record(data_item, sep=separator)
+
+        flat_json_data.append(flat_data_item)
+
+    return flat_json_data
+
+
+def update_row_columns(fields, data):
+    """
+    Set missing column values
+    :param fields: table fields
+    :param data: data to be dumped
+    :return data: updated data
+    """
+    columns = [item.get('name') for item in fields]
+    for row_data in list(data):
+        row_columns = row_data.keys()
+
+        missing_columns = list(set(columns) - set(row_columns))
+
+        if len(missing_columns) > 0:
+            for column in missing_columns:
+                field_obj = next(
+                    filter(
+                        lambda field: field.get('name') == column,
+                        fields
+                    )
+                )
+                row_data.setdefault(
+                    column,
+                    set_column_defaults(field_obj.get('type', None))
+                )
+    return data
+
+
+def set_column_defaults(type):
+    """
+    Set data column default values for missing columns
+    : param type: data type
+    : return column_string: Postgres query compatible string
+    """
+    if type.lower() == 'int':
+        return None
+
+    if type.lower() == 'decimal':
+        return None
+
+    if type.lower() == 'char':
+        return ''
+
+    if type.lower() == 'boolean':
+        return None
+
+    else:
+        return ''
+
+
+def clean_key_field(row, primary_key):
+    """
+    strip off string 'uuid:' from key field
+    :param row: the row object
+    :param primary_key: primary key
+    :return row: row object with cleaned key field
+    """
+    row[primary_key] = row.get(primary_key).replace('uuid:', '').strip()
+
+    return row
+
+
+############################
+#  Postgres Operations     #
+############################
+def construct_postgres_create_table_query(table_name, columns_data):
+    """
+    construct postgres CREATE TABLE IF NOT EXISTS query
+    :param table_name: name of the table to create
+    :param columns_data: the data column names
+    :return create_table_query: SQL query string
+    """
+    create_table_query = 'CREATE TABLE IF NOT EXISTS ' \
+                         + table_name + ' (' + ', '.join(columns_data) + ')'
+
+    return create_table_query
+
+
+def construct_postgres_upsert_query(table_name, columns, target_column):
+    """
+    construct INSERT or UPDATE (UPSERT) query for postgres
+    :param table_name: the table to upsert its row(s)
+    :param columns: the table column names
+    :param target_column: reference column for update
+    :return full_upsert_query_string:  complete UPSERT query string
+    """
+    insert_query_string = 'INSERT INTO ' + table_name + '(' + ','\
```

Comment:
Is there a way we could build these queries using a library like `SQLAlchemy` instead of building them manually like this?

Reply:
Am avoiding being boxed in away of doing this. Optimizing raw queries is always much simpler than using ORMs. Flexibility is the  point.